File: csharp/Oojs/TypeDef.cs
namespace Oojs;

public sealed class TypeDef
{
    public TypeDef(
        string name,
        string schemaId,
        bool isAbstract = false,
        string? extends = null,
        string? discriminatorValue = null,
        string title = "",
        string description = "")
    {
        Name = name;
        SchemaId = schemaId;
        Abstract = isAbstract;
        Extends = extends;
        DiscriminatorValue = discriminatorValue;
        Title = title;
        Description = description;
    }

    public string Name { get; }
    public string SchemaId { get; }
    public bool Abstract { get; }
    public string? Extends { get; }
    public string? DiscriminatorValue { get; }
    public string Title { get; }
    public string Description { get; }

    public Dictionary<string, PropertyDef> OwnProperties { get; } = new();
    public List<string> OwnRequired { get; } = new();
    public TypeDef? Supertype { get; set; }

    public string EffectiveDiscriminatorValue => DiscriminatorValue ?? Name;

    public List<string> EffectiveRequired()
    {
        if (Supertype is null)
        {
            return new List<string>(OwnRequired);
        }
        var req = Supertype.EffectiveRequired();
        req.AddRange(OwnRequired);
        return req;
    }

    public Dictionary<string, PropertyDef> EffectiveProperties()
    {
        if (Supertype is null)
        {
            return new Dictionary<string, PropertyDef>(OwnProperties);
        }
        var props = Supertype.EffectiveProperties();
        foreach (var kvp in OwnProperties)
        {
            props[kvp.Key] = kvp.Value;
        }
        return props;
    }

    public bool IsSubtypeOf(TypeDef other)
    {
        var current = this as TypeDef?;
        while (current is not null)
        {
            if (ReferenceEquals(current, other))
            {
                return true;
            }
            current = current.Supertype;
        }
        return false;
    }
}

File: php/src/Model/TypeDef.php
<?php

declare(strict_types=1);

namespace Oojs\Model;

/**
 * An OOJS type definition.
 *
 * Immutable fields are set via the constructor.
 * Mutable fields (ownProperties, ownRequired, supertype) are populated by the
 * Registry during loading and should not be modified afterwards.
 */
class TypeDef
{
    /** @var array<string, PropertyDef> Own (non-inherited) properties */
    public array $ownProperties = [];

    /** @var list<string> Own (non-inherited) required property names */
    public array $ownRequired = [];

    /** Resolved supertype, set by Registry during hierarchy resolution. */
    public ?TypeDef $supertype = null;

    public function __construct(
        public readonly string $name,
        public readonly string $schemaId,
        public readonly bool $abstract = false,
        public readonly ?string $extends = null,
        public readonly ?string $discriminatorValue = null,
        public readonly string $title = '',
        public readonly string $description = '',
    ) {}

    /**
     * The discriminator value used in instances.
     * Falls back to the type name if no override is declared.
     */
    public function getEffectiveDiscriminatorValue(): string
    {
        return $this->discriminatorValue ?? $this->name;
    }

    /**
     * Union of all required property names in the inheritance chain,
     * from root to self (ancestor-first order).
     *
     * @return list<string>
     */
    public function effectiveRequired(): array
    {
        if ($this->supertype === null) {
            return $this->ownRequired;
        }
        return array_merge($this->supertype->effectiveRequired(), $this->ownRequired);
    }

    /**
     * Merged property map from the full inheritance chain.
     * Child properties shadow ancestor properties with the same name (though
     * redeclaration is forbidden at load time, so no shadowing occurs in
     * valid schemas).
     *
     * @return array<string, PropertyDef>
     */
    public function effectiveProperties(): array
    {
        if ($this->supertype === null) {
            return $this->ownProperties;
        }
        return array_merge($this->supertype->effectiveProperties(), $this->ownProperties);
    }

    /**
     * Returns true if $this equals $other or $this descends from $other.
     */
    public function isSubtypeOf(TypeDef $other): bool
    {
        $current = $this;
        while ($current !== null) {
            if ($current === $other) {
                return true;
            }
            $current = $current->supertype;
        }
        return false;
    }
}
